addons/binding/org.openhab.binding.fritzaha/src/main/java/org/openhab/binding/fritzaha/internal/discovery/FritzBoxDiscoveryService.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.fritzaha.internal.discovery;
+
+import static org.openhab.binding.fritzaha.FritzAhaBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.fritzaha.handler.FritzAhaBridgeHandler;
+import org.openhab.binding.fritzaha.handler.FritzAhaHandler;
+import org.openhab.binding.fritzaha.internal.model.DeviceModel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Discover all AHA (AVM Home Automation) devices connected to a FRITZ!Box
+ * device.
+ *
+ * @author Robert Bausdorf
+ * @since 2.0
+ */
+public class FritzBoxDiscoveryService extends AbstractDiscoveryService {
+
+	/**
+	 * Logger
+	 */
+	private final Logger LOG = LoggerFactory
+			.getLogger(FritzBoxDiscoveryService.class);
+
+	/**
+	 * Maximum time to search for devices.
+	 */
+	private final static int SEARCH_TIME = 30;
+
+	/**
+	 * Handler of the bridge of which devices have to be discovered.
+	 */
+	private FritzAhaBridgeHandler bridgeHandler;
+
+	public FritzBoxDiscoveryService(FritzAhaBridgeHandler bridgeHandler) {
+		super(SEARCH_TIME);
+		this.bridgeHandler = bridgeHandler;
+		if (bridgeHandler == null) {
+			LOG.warn("no bridge handler for scan given");
+		}
+	}
+
+	/**
+	 * Provide supported thing types.
+	 * 
+	 * @return Set of supported thing type UIDs
+	 */
+	@Override
+	public Set<ThingTypeUID> getSupportedThingTypes() {
```

Comment:
you can remove this method when you pass the supportedTypes in the constructor (see above)